src/js/node/events.js

```diff
@@ -386,21 +388,20 @@ async function* on(emitter, event, options = {}) {
 
   for (const evName of options?.close || []) {
     emitter.on(evName, () => {
-      emitter.emit(event, undefined);
       done = true;
     });
   }
 
-  while (!done) {
+  while (!done || !unconsumedEvents.isEmpty()) {
     if (!unconsumedEvents.isEmpty()) {
       yield Promise.$resolve(unconsumedEvents.shift());
-    }
-    if (!unconsumedErrors.isEmpty()) {
+    } else if (!unconsumedErrors.isEmpty()) {
       yield Promise.$reject(unconsumedErrors.shift());
+    } else {
```

Comment:
The other `yield` statements needed to be enclosed by `else` because otherwise when .next() is called on the iterator in the `for await...of` loop, the generator function started at the wrong position.

E.g. following situation:
- Pre: Iterator is created via `events.on`, and shortly after 3 events are emitted at the next tick.
1. `.next()` call: No events or errors are queued, so a new unresolved promise is returned. When the first event is emitted in the next tick, the promise is resolved and `.next()` is called. The other two events are put into `unconsumedEvents`.
2. Call: `unconsumedEvents` does have items, so the first yield statement is called.
3. Call: The execution continues after the first yield statement. Since the other yields were not enclosed by `else` a new `unconsumedPromise` is created and yielded. Even though there is a `unconsumedEvent` available.

This last promise is never resolved because of that. This only happens when 3+ events are emitted, so the test case did not catch that.